ruby/ext/google/protobuf_c/encode_decode.c

```diff
@@ -174,9 +174,220 @@ static void *submsg_handler(void *closure, const void *hd) {
   return submsg;
 }
 
+// Handler data for startmap/endmap handlers.
+typedef struct {
+  size_t ofs;
+  upb_fieldtype_t key_field_type;
+  upb_fieldtype_t value_field_type;
+  VALUE value_field_typeclass;
```

Comment:
Maybe I spoke too soon -- is it the case that we always mark a descriptor that marks this?

I leave it to you whether it's better to have typeclass_references or to rely on a proof of this sort. If we do the latter there should be a quick comment describing how we know this is rooted.